### PR TITLE
feat(profiling): move flamegraph and differential flamegraph

### PR DESCRIPTION
**Pull request by JonasBa**
Port over the main flamegraph and differential flamegraph classes.

The two classes are used to build (not rendering) the flamegraph in different ways (see buildCallOrderGraph and buildLeftHeavyGraph). They are mainly responsible for outputting an array of frames to render and giving us a configuration view (usually at x:0, y:0, width: timeline duration, height: max stack height).

There are two main functions to look at - buildLeftHeavyGraph and buildCallOrderGraph.
They both iterate over all frames of a profile and append the frames to render to `this.frames` with the slight difference that buildLeftHeavyGraph sorts the tree first (depth first) and updates the frames startTime and endTime based on their new positions so that startTime is the sum of all child frames preceding the current frame. 

<details>
<summary>e.g. left heavy timings change</summary>
CleanShot 2022-01-13 at 10 19 28@2x
</details>

Note that the order of this.frames doesn't matter and it will be up to whoever consumes the flamegraph to determine what and when to render.

### CODE REVIEW

File: static/app/utils/profiling/differentialFlamegraph.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import {FlamegraphTheme} from './flamegraph/FlamegraphTheme';
+import {relativeChange} from './units/units';
+import {Flamegraph} from './flamegraph';
+import {FlamegraphFrame} from './flamegraphFrame';
+
+function countFrameOccurences(frames: FlamegraphFrame[]): Map<string, number> {
+  const counts = new Map<string, number>();
+
+  for (const frame of frames) {
+    const key = frame.frame.name + (frame.frame.file ? frame.frame.file : '');
```
Comment (Zylphrex) [inline static/app/utils/profiling/differentialFlamegraph.tsx:10]: Should this be a method on `Frame`?

File: static/app/utils/profiling/flamegraph.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+    this.frames = leftHeavy
+      ? this.buildLeftHeavyGraph(profile)
+      : this.buildCallOrderGraph(profile);
+
+    if (this.frames.length) {
+      this.configSpace = new Rect(0, 0, this.duration, this.depth);
+    } else {
+      // If we have no frames, set the trace duration to 1 second so that we can render a placeholder grid
+      this.configSpace = new Rect(0, 0, 1_000_000, 0);
```
Comment (Zylphrex) [inline static/app/utils/profiling/flamegraph.ts:53]: Are the units always going to be `microseconds`? I see that it's possible to specify a unit in `Profile`. Should that be used to compute this value?

**Review by Zylphrex — APPROVED** [review]

File: static/app/utils/profiling/differentialFlamegraph.tsx (hunk 1 above)
Comment (JonasBa) [reply static/app/utils/profiling/differentialFlamegraph.tsx:10]: @Zylphrex I thought about it before, but I haven't done it. The reason for that is that this is more the key that we use for color encoding which is not something that the Frame has to know or care about. Maybe a utility fn would be a good middle ground here, wdyt?

File: static/app/utils/profiling/flamegraph.ts (hunk 2 above)
Comment (JonasBa) [reply static/app/utils/profiling/flamegraph.ts:53]: Ah correct, we need to set that too, I missed that 🙏🏼

**Pull request merged**